papers/hatra21/hatra21.tex

```diff
@@ -325,15 +338,62 @@ \subsection{Mixing types}
 \item Can we have strict and non-strict mode infer the same types,
   only with different flagging?
 
+\item Is strict-mode code sound when it relies on non-strict code,
+  which has weaker invariants?
+
 \end{itemize}
 \emph{Related work}: this appears to be an under-explored area.
 
+\subsection{Type inference}
+
+Goal: \emph{infer types to allow gradual adoption of type annotations.}
+
+Roblox, for many years, used the Lua language, which is dynamically typed and possesses a
+very weak type system. Due to this large quantity of pre-existing dynamically-typed code,
+it is essential for the type system to function even in the absence of type annotations,
+in order to make the type features of Luau gradually adoptable. This means that Luau needs
+to be able to infer types for symbols without any annotations being present, to the best of
+its ability. This precludes syntactical rules such as Rust's requirement that all function
+parameters be annotated with their type~\cite[Ch. 3.3]{RustBook}.
+
+This requirement presents challenges for the type inference algorithm, because Luau may not have
+enough information to determine the type of a given program. In non-strict mode in particular,
+which sees use in existing codebases, we cannot rely on the presence of type annotations. We also
+cannot require that the user provide them if Luau cannot deduce the type of a symbol. In cases
+such as this, we must admit defeat and assume that the code is correct, to fulfill non-strict
+mode's goal of ``no false positives''. We do this by saying that the result of the operation is
+$\mathsf{any}$, a type that can be converted to and from any other type freely.
+
+In strict mode, Luau is not so limited, and in pursuit of the strict-mode goal of ``no false
+negatives'', we may surface errors to the user indicating that the type inference system requires
+more information, in the form of annotations, in order to type-check a piece of code. This code,
+for example, requires a type annotation in order for Luau to determine the return type of the function:
```

Comment:
For people not familiar with Luau, it's worth spelling out why.
```suggestion
for example, requires a type annotation in order for Luau to determine the return type of the function (since Luau does not know if ``+'' refers to built-in addition on numbers, or a user-defined method):
```